File: u2leap.mjs
/**
 * Unicode Telugu -> TL-Hemalatha (iLEAP) converter.
 *
 * Replicates the logic of the legacy VB6 "UNI 2 LEAP" tool, reconstructed from
 * its data files (U2LEAP.dll = glyph table, UNI_DECIM.dll = token table).
 *
 * Glyph table layout (indices into U2LEAP):
 *   1..19    independent vowels + candrabindu/anusvara/visarga
 *   20..559  36 consonants (35 + ksha) x 15 forms, stride 36:
 *            base(20) aa(56) i(92) ii(128) u(164) uu(200) r.(236) rr.(272)
 *            e(308) ee(344) ai(380) o(416) oo(452) au(488) halant(524)
 *   700..734 vattu (subjoined consonant) glyphs, same consonant order
 *   735..777 specials: alt ra-vattu(737), digits, quotes, ksha series(758..772)
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const tables = JSON.parse(
  readFileSync(join(dirname(fileURLToPath(import.meta.url)), 'tables.json'), 'utf8'),
);
const g = (i) => tables.glyphs[String(i)] ?? '';

const CONSONANTS = [...'కఖగఘఙచఛజఝఞటఠడఢణతథదధనపఫబభమయరఱలళవశషసహ'];
const RANK = new Map(CONSONANTS.map((c, i) => [c, i]));

const BASE_BLOCK = 20;
const HALANT_BLOCK = 524;
const VATTU_BASE = 700;
const ALT_RA_VATTU = 737; // 'û' — used for ra in 3-consonant clusters (స్త్ర)

const MATRA_BLOCK = new Map([
  ['ా', 56], ['ి', 92], ['ీ', 128], ['ు', 164], ['ూ', 200],
  ['ృ', 236], ['ౄ', 272], ['ె', 308], ['ే', 344], ['ై', 380],
  ['ొ', 416], ['ో', 452], ['ౌ', 488],
]);

// క్ష ligature series (glyph indices 758..772)
const KSHA_SERIES = new Map([
  ['', 758], ['ా', 759], ['ి', 760], ['ీ', 761], ['ు', 762], ['ూ', 763],
  ['ృ', 764], ['ౄ', 765], ['ె', 766], ['ే', 767], ['ై', 768],
  ['ొ', 769], ['ో', 770], ['ౌ', 771], ['్', 772],
]);

const SINGLES = new Map(
  [...'అఆఇఈఉఊఋౠఎఏఐఒఓఔఌౡఁంః'].map((c, i) => [c, i + 1]),
);

const TELUGU_DIGITS = new Map([...'౦౧౨౩౪౫౬౭౮౯'].map((c, i) => [c, String(i)]));

// From the token table: legacy app's own substitutions for non-Telugu chars.
const MISC = new Map([
  ['₹', g(777)],          // 'Rs'
  ['‘', g(752)], ['’', g(752)],   // -> '
  ['“', g(751)], ['”', g(751)],   // -> "
  ['…', g(776)],          // '...'
  ['​', ''], ['‌', ''], ['‍', ''], // ZW chars consumed
]);

const VIRAMA = '్';
const ZWNJ = '‌';
const ZWJ = '‍';

// Rare letters approximated to their nearest classical equivalent
const APPROX = new Map([['ౘ', 'చ'], ['ౙ', 'జ'], ['ౚ', 'ఱ'], ['ఴ', 'ళ']]);
// Combining marks/signs with no Hemalatha equivalent: dropped like stray matras
const DROP = new Set([...'ౕౖౢౣఀఄ఼ఽ౷']);

const isConsonant = (c) => RANK.has(c);

/**
 * Convert Unicode Telugu text to TL-Hemalatha glyph text.
 * Output chars map 1:1 to the legacy font's byte values via latin-1.
 */
export function convert(input) {
  const text = [...input.normalize('NFC')]
    .filter((c) => !DROP.has(c))
    .map((c) => APPROX.get(c) ?? c);
  let out = '';
  let i = 0;

  while (i < text.length) {
    const c = text[i];

    if (isConsonant(c)) {
      // --- parse one cluster: C (virama C)* (matra | virama-final)? ---
      let base = c;
      const subs = [];
      let j = i + 1;
      let halantFinal = false;
      while (text[j] === VIRAMA) {
        const nxt = text[j + 1];
        if (nxt !== undefined && isConsonant(nxt)) {
          // virama + consonant = subjoined vattu
          subs.push(nxt);
          j += 2;
        } else if (nxt === ZWNJ || nxt === ZWJ) {
          halantFinal = true;
          j += 2;
          break;
        } else {
          halantFinal = true;
          j += 1;
          break;
        }
      }
      let matra = '';
      if (!halantFinal && MATRA_BLOCK.has(text[j])) {
        matra = text[j];
        j += 1;
      }

      // --- emit ---
      if (base === 'క' && subs[0] === 'ష') {
        // క్ష ligature takes the whole matra/halant slot
        out += g(KSHA_SERIES.get(halantFinal ? '్' : matra));
        out += vattus(subs.slice(1));
      } else {
        const r = RANK.get(base);
        if (halantFinal) out += g(HALANT_BLOCK + r);
        else if (matra) out += g(MATRA_BLOCK.get(matra) + r);
        else out += g(BASE_BLOCK + r);
        out += vattus(subs);
      }
      i = j;
      continue;
    }

    if (SINGLES.has(c)) { out += g(SINGLES.get(c)); i += 1; continue; }
    if (TELUGU_DIGITS.has(c)) { out += TELUGU_DIGITS.get(c); i += 1; continue; }
    if (MISC.has(c)) { out += MISC.get(c); i += 1; continue; }
    if (MATRA_BLOCK.has(c) || c === VIRAMA) { i += 1; continue; } // stray matra: drop
    // pass through, but never let a >0xFF char corrupt the byte output
    out += c.codePointAt(0) <= 0xff ? c : '?';
    i += 1;
  }
  return out;
}

function vattus(subs) {
  let s = '';
  for (let k = 0; k < subs.length; k += 1) {
    const c = subs[k];
    if (c === 'ర' && k > 0) s += g(ALT_RA_VATTU); // స్త్ర-style alt ra-vattu
    else s += g(VATTU_BASE + RANK.get(c));
  }
  return s;
}

/** Encode converted text to the raw legacy byte values. */
export function toBytes(hemalatha) {
  return Buffer.from(hemalatha, 'latin1');
}

// --- CLI: node u2leap.mjs "తెలుగు"  (or pipe stdin) ---
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const arg = process.argv[2];
  const input = arg ?? readFileSync(0, 'utf8');
  process.stdout.write(convert(input) + (arg ? '\n' : ''));
}

File: u2leap.py
"""Unicode Telugu -> TL-Hemalatha (iLEAP) converter.

Python port of u2leap.mjs — same tables.json, same algorithm.
See README.md for the table layout and algorithm details.
"""
import json
import unicodedata
from pathlib import Path

_tables = json.loads((Path(__file__).parent / "tables.json").read_text(encoding="utf-8"))
_GLYPHS = _tables["glyphs"]


def _g(i: int) -> str:
    return _GLYPHS.get(str(i), "")


CONSONANTS = list("కఖగఘఙచఛజఝఞటఠడఢణతథదధనపఫబభమయరఱలళవశషసహ")
RANK = {c: i for i, c in enumerate(CONSONANTS)}

BASE_BLOCK = 20
HALANT_BLOCK = 524
VATTU_BASE = 700
RA_HOOK = 726       # 'ú' — pre-positioned ్ర hook, emitted BEFORE the base glyph
ALT_RA_VATTU = 737  # 'û' — ్ర after another vattu (స్త్ర pattern), appended after
# Glyphs whose vattu is not at VATTU_BASE + rank (confirmed against known-good
# output): ్ట is the plain-cup glyph at 727, not the loop+cup at 710.
VATTU_OVERRIDES = {"ట": 727}

MATRA_BLOCK = {
    "ా": 56, "ి": 92, "ీ": 128, "ు": 164, "ూ": 200,
    "ృ": 236, "ౄ": 272, "ె": 308, "ే": 344, "ై": 380,
    "ొ": 416, "ో": 452, "ౌ": 488,
}

KSHA_SERIES = {
    "": 758, "ా": 759, "ి": 760, "ీ": 761, "ు": 762, "ూ": 763,
    "ృ": 764, "ౄ": 765, "ె": 766, "ే": 767, "ై": 768,
    "ొ": 769, "ో": 770, "ౌ": 771, "్": 772,
}

SINGLES = {c: i + 1 for i, c in enumerate("అఆఇఈఉఊఋౠఎఏఐఒఓఔఌౡఁంః")}
TELUGU_DIGITS = {c: str(i) for i, c in enumerate("౦౧౨౩౪౫౬౭౮౯")}

MISC = {
    "₹": _g(777),                       # 'Rs'
    "‘": _g(752), "’": _g(752),  # ‘ ’ -> '
    "“": _g(751), "”": _g(751),  # “ ” -> "
    "…": _g(776),                  # … -> '...'
    "​": "", "‌": "", "‍": "",
}

VIRAMA = "్"
ZWNJ = "‌"
ZWJ = "‍"

# Rare letters approximated to their nearest classical equivalent
APPROX = {"ౘ": "చ", "ౙ": "జ", "ౚ": "ఱ", "ఴ": "ళ"}
# Combining marks/signs with no Hemalatha equivalent: dropped like stray matras
DROP = set("ౕౖౢౣఀఄ఼ఽ౷")


def _vattus(subs: list, after_vattu: bool = False) -> str:
    out = ""
    for k, c in enumerate(subs):
        if c == "ర" and (k > 0 or after_vattu):
            out += _g(ALT_RA_VATTU)
        elif c in VATTU_OVERRIDES:
            out += _g(VATTU_OVERRIDES[c])
        else:
            out += _g(VATTU_BASE + RANK[c])
    return out


def convert(text: str) -> str:
    """Convert Unicode Telugu to TL-Hemalatha glyph text.

    Output chars map 1:1 to the legacy font's byte values via latin-1.
    """
    chars = [
        APPROX.get(c, c)
        for c in unicodedata.normalize("NFC", text)
        if c not in DROP
    ]
    out = []
    i, n = 0, len(chars)

    while i < n:
        c = chars[i]

        if c in RANK:
            base = c
            subs = []
            j = i + 1
            halant_final = False
            while j < n and chars[j] == VIRAMA:
                nxt = chars[j + 1] if j + 1 < n else None
                if nxt is not None and nxt in RANK:
                    subs.append(nxt)
                    j += 2
                elif nxt in (ZWNJ, ZWJ):
                    halant_final = True
                    j += 2
                    break
                else:
                    halant_final = True
                    j += 1
                    break
            matra = ""
            if not halant_final and j < n and chars[j] in MATRA_BLOCK:
                matra = chars[j]
                j += 1

            # ్ర in a cluster: pre-positioned hook ú before the whole cluster
            # glyph (ప్రై = ú\|ms, స్ట్రీ = ú{qsí). Exception: the ్త్ర family
            # (స్త్ర/స్త్రీ) keeps the legacy post-form ò+û per tokens 133/193.
            tra_pair = len(subs) >= 2 and subs[-1] == "ర" and subs[-2] == "త"
            pre_ra = "ర" in subs and not tra_pair
            if pre_ra:
                out.append(_g(RA_HOOK))
                subs = list(subs)
                subs.remove("ర")
            if base == "క" and subs and subs[0] == "ష":
                out.append(_g(KSHA_SERIES["్" if halant_final else matra]))
                out.append(_vattus(subs[1:], after_vattu=True))
            else:
                r = RANK[base]
                if halant_final:
                    out.append(_g(HALANT_BLOCK + r))
                elif matra:
                    out.append(_g(MATRA_BLOCK[matra] + r))
                else:
                    out.append(_g(BASE_BLOCK + r))
                out.append(_vattus(subs, after_vattu=pre_ra))
            i = j
            continue

        if c in SINGLES:
            out.append(_g(SINGLES[c]))
        elif c in TELUGU_DIGITS:
            out.append(TELUGU_DIGITS[c])
        elif c in MISC:
            out.append(MISC[c])
        elif c in MATRA_BLOCK or c == VIRAMA:
            pass  # stray matra: drop
        else:
            # pass through, but never let a >0xFF char corrupt the byte output
            out.append(c if ord(c) <= 0xFF else "?")
        i += 1

    return "".join(out)


def to_bytes(hemalatha: str) -> bytes:
    """Raw legacy byte values (latin-1)."""
    return hemalatha.encode("latin-1")
